### PR TITLE
[DOC] Reshape feature

### PR DESCRIPTION
**Pull request by jane-intel**

### CODE REVIEW

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

File: docs/IE_DG/ShapeInference.md [hunk 1]
```diff
@@ -1,22 +1,45 @@
...
-To feed input data of a shape that is different from the model input shape, resize the model first.
+Inference Engine takes three kinds of model description as an input, which are all converted to `InferenceEngine::CNNNetwork` object:
+1. [Intermediate Representation (IR)](../MO_DG/IR_and_opsets.md) through `InferenceEngine::Core::ReadNetwork`
+2. [ONNX model](../IE_DG/OnnxImporterTutorial.md) through `InferenceEngine::Core::ReadNetwork`
+3. [nGraph::Function](../IE_DG/nGraph_Flow.md) through constructor of `InferenceEngine::CNNNetwork`
 
-Model resizing on the stage of <a href="_docs_MO_DG_prepare_model_convert_model_Converting_Model_General.html#when_to_specify_input_shapes">IR generation</a> or [nGraph::Function creation](nGraphTutorial.md) is the recommended approach. 
-OpenVINO™ provides the following experimental methods for runtime model reshaping:
+`InferenceEngine::CNNNetwork` keeps `ngraph::Function` object with the model description.
+It should have static (fixed) input shapes to be successfully loaded to the Inference Engine plugins.
```
[the remark this answers is not in the document]
Comment (jane-intel) [reply docs/IE_DG/ShapeInference.md]: I got rid of words `static` and `dynamic` in favor to `(not) fully defined` as we use it starting from Model Optimizer error messages. Regarding the reference to a document -- I failed to find one in our documentation. Probably it is placed in some nGraph specific place. But I believe our users are already familiar with this terms. What do you think?

**Review by lazarevevgeny — APPROVED** [review]

**Review by andrew-zaytsev — DISMISSED** [review]

File: docs/IE_DG/ShapeInference.md [hunk 2]
```diff
@@ -1,38 +1,60 @@
...
+   The `InferenceEngine::CNNNetwork::reshape` method updates input shapes and propagates them down to the outputs of the model through all intermediate layers. It is allowed to reshape model multiple times like in this application scheme:
+   ```
+   ReadNetwork -> reshape(input_1_shape) -> LoadNetwork -> infer(input_1)
+              \
+               -> reshape(input_2_shape) -> LoadNetwork -> infer(input_2)
+   ```
+   > **NOTES**:
+   > - Starting with the 2021.1 release, the Model Optimizer converts topologies keeping shape-calculating sub-graphs by default, which allows correct shape propagation during    reshaping.<br>
+   > - Older versions of IRs are not guaranteed to reshape successfully. Please re-generate them with the Model Optimizer of the latest version of OpenVINO™.<br>
+   > - It is required to reshape the network before loading to the plugin for ONNX models with not fully defined input shape that were imported with the ONNX importer.
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:38]: Sounds a bit confusing. I tried rephrasing, is this what is meant?

```
If an ONNX model does not have a fully defined input shape and the model was imported with the ONNX importer, reshape the model before loading it to the plugin.
```

File: docs/IE_DG/ShapeInference.md [hunk 3]
```diff
@@ -1,38 +1,60 @@
...
+* **Set a new input shape** with the `InferenceEngine::CNNNetwork::reshape` method.<br>
+   The `InferenceEngine::CNNNetwork::reshape` method updates input shapes and propagates them down to the outputs of the model through all intermediate layers. It is allowed to reshape model multiple times like in this application scheme:
+   ```
+   ReadNetwork -> reshape(input_1_shape) -> LoadNetwork -> infer(input_1)
+              \
+               -> reshape(input_2_shape) -> LoadNetwork -> infer(input_2)
+   ```
+   > **NOTES**:
+   > - Starting with the 2021.1 release, the Model Optimizer converts topologies keeping shape-calculating sub-graphs by default, which allows correct shape propagation during    reshaping.<br>
+   > - Older versions of IRs are not guaranteed to reshape successfully. Please re-generate them with the Model Optimizer of the latest version of OpenVINO™.<br>
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:37]: `regenerate`

File: docs/IE_DG/ShapeInference.md [hunk 4]
```diff
@@ -1,38 +1,60 @@
...
+   > - It is required to reshape the network before loading to the plugin for ONNX models with not fully defined input shape that were imported with the ONNX importer.
+* **Set a new batch dimension value** with the `InferenceEngine::CNNNetwork::setBatchSize` method.<br>     
+   The meaning of a model batch may vary depending on choices you made during the model designing.
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:40]: --> ```The meaning of a model batch may vary depending on the model design. ```

File: docs/IE_DG/ShapeInference.md [hunk 5]
```diff
@@ -1,38 +1,60 @@
...
+   ```
+   > **NOTES**:
+   > - Starting with the 2021.1 release, the Model Optimizer converts topologies keeping shape-calculating sub-graphs by default, which allows correct shape propagation during    reshaping.<br>
+   > - Older versions of IRs are not guaranteed to reshape successfully. Please re-generate them with the Model Optimizer of the latest version of OpenVINO™.<br>
+   > - It is required to reshape the network before loading to the plugin for ONNX models with not fully defined input shape that were imported with the ONNX importer.
+* **Set a new batch dimension value** with the `InferenceEngine::CNNNetwork::setBatchSize` method.<br>     
+   The meaning of a model batch may vary depending on choices you made during the model designing. 
+   The `InferenceEngine::CNNNetwork::setBatchSize` method deduces index of batch dimension relying only on the input rank. 
+   This method does not work for models with a non-zero index batch placement or models with inputs without a batch dimension. 
+   Batch-setting algorithm does not involve shape inference mechanism.
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:43]: The algorithm, the mechanism

File: docs/IE_DG/ShapeInference.md [hunk 6]
```diff
@@ -1,38 +1,60 @@
...
+   Batch of input and output shapes for all layers is set to a new batch value without layer validation.
+   It may cause both positive and negative side effects.
+   Due to the limitations described above, the current method is not recommended to use.
+   If there is a need to set new batch size for the model, please use `CNNNetwork::reshape` method instead.
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:47]: --> ```If you need to set a new batch size for the model, use the `CNNNetwork::reshape` method instead.```

File: docs/IE_DG/ShapeInference.md [hunk 7]
```diff
@@ -1,38 +1,60 @@
...
+
+Please do not use runtime reshaping methods together, especially do not call `CNNNetwork::reshape` method after `InferenceEngine::CNNNetwork::setBatchSize` usage.
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:49]: --> 
```Do not use runtime reshaping methods simultaneously, especially do not call the `CNNNetwork::reshape` method after  you use `InferenceEngine::CNNNetwork::setBatchSize`.```

File: docs/IE_DG/ShapeInference.md [hunk 8]
```diff
@@ -1,38 +1,60 @@
...
+`InferenceEngine::CNNNetwork::setBatchSize` method causes irreversible conversion of internal model representation to the legacy model representation.
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:50]: --> 
```
The `InferenceEngine::CNNNetwork::setBatchSize` method causes irreversible conversion of the internal model representation into the legacy model representation.
```

File: docs/IE_DG/ShapeInference.md [hunk 9]
```diff
@@ -1,38 +1,60 @@
...
+It does not use `nGraph` for shape inference which lead to reduced reshape opportunities and may affect performance of the model.
```
Comment (aalborov) [inline docs/IE_DG/ShapeInference.md:51]: The method does not use `nGraph` for shape inference which leads to reduced reshape opportunities and may affect the performance of the model.

**Review by aalborov — CHANGES_REQUESTED** [review]

File: docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md [hunk 10]
```diff
@@ -1,10 +1,9 @@
 # Converting TensorFlow* Object Detection API Models {#openvino_docs_MO_DG_prepare_model_convert_model_tf_specific_Convert_Object_Detection_API_Models}
 
 > **NOTES**:
->
-> * Starting with the 2019 R1 release, the Model Optimizer supports the `--keep_shape_ops` command line parameter that allows you to convert the TensorFlow\* Object Detection API Faster and Mask RCNNs topologies so they can be re-shaped in the Inference Engine using dedicated reshape API. Refer to [Using Shape Inference](../../../../IE_DG/ShapeInference.md) for more information on how to use this feature. It is possible to change the both spatial dimensions of the input image and batch size.
 > * Starting with the 2018 R4 release, the Model Optimizer supports the `--input_shape` command line parameter for the TensorFlow\* Object Detection API topologies. Refer to the [Custom Input Shape](#tf_od_custom_input_shape) for more information.
```
[the remark this answers is not in the document]
Comment (jane-intel) [reply docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md:4]: @lazarevevgeny, do we?

[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply docs/MO_DG/prepare_model/convert_model/tf_specific/Convert_Object_Detection_API_Models.md:5]: Let's remove them.

**Review by aalborov — APPROVED** [review]
LGTM

**Pull request merged**